### ISSUE TITLE
Cups_queue defaults not used on Ubuntu 16.04

### ISSUE THREAD

**Issue by imphil** [leoarnold/puppet-cups#5]
**Given** Ubuntu 16.04 with Ruby 2.3.0, Puppet 3.8.5 and CUPS 2.1.3

**When** I apply the manifest

```puppet
include '::cups'

  # default cups settings
  Cups_queue {
    accepting  => true,
    enabled    => true,
  }

  cups_queue { 'name':
    ensure      => 'printer',
    uri         => 'socket://url:9100',
    description => 'xxx',
    model       => 'postscript-hp:0/ppd/hplip/HP/hp-laserjet_400_m401dne-ps.ppd',
    location    => 'xxx',
  }
```

the printer is set to enabled = false/accepting = false. When I add enabled/accepting to the cups_queue instance (not to the default), the queues are accepting and enabled. So it seems the default values are not taken into account.

Do you need any additional logs for that? Which ones?

**Comment by leoarnold** [comment]
This is certainly unexpected. Please delete the queue using `lpadmin -x name` and then rerun your above manifest using `puppet apply -dv manifest.pp &> issue.txt` and attach the created `issue.txt` to your answer.

Please do **not paste** the log file, as it will be very long ;-)

**Comment by leoarnold** [comment]
Might it be that your log features the line

```
Error: Execution of '/usr/sbin/lpadmin -E -p Testo -m postscript-hp:0/ppd/hplip/HP/hp-laserjet_400_m401dne-ps.ppd' returned 1: lpadmin: Unable to copy PPD file.
```

because the package `hplip` is absent? If so, please try

```Puppet
package { 'hplip':
  ensure => 'present',
}

class { '::cups':
  require => Package['hplip'],
}

Cups_queue {
  accepting  => true,
  enabled    => true,
}

cups_queue { 'name':
  ensure      => 'printer',
  uri         => 'socket://url:9100',
  description => 'xxx',
  model       => 'postscript-hp:0/ppd/hplip/HP/hp-laserjet_400_m401dne-ps.ppd',
  location    => 'xxx',
}

**Comment by imphil** [comment]
Hi,

thanks for looking into it. It turns out, this was a bug on my side. I missed that the resource defaults only apply within the scope they are declared in. I had two classes, one setting up all the basics (which also contained the Cups_queue defaults), and another one which contained the printers itself. In this case, the defaults were not taken. Moving the Cups_queue defaults to global scope (i.e. out of the class definition) works as expected.

With that in mind, I had a look at the documentation again, and while it's mentioned there (`# Default values for all 'cups_queue' resources in the current scope:`), maybe you can highlight that a bit more prominently? I guess others following the example in "Recommended refactor" might run into the same issue.

**Comment by imphil** [comment]
Looks like I was to fast to jump to conclusions here. Setting `Cups_queue` with default in global scope does not yield the expected results, it is ignored. So the result is that I need to add `Cups_queue` with the defaults in all classes that use `cups_queue`. Or am I missing something?

**Comment by leoarnold** [comment]
`Cups_queue` (a so-called *resource default*) is not implemented by the module, but provided by Puppet itself. Please see the Puppet documentation on [resource defaults](https://docs.puppet.com/puppet/3.8/reference/lang_defaults.html), in particular the subsection on its *area of effect*.

I'll amend the module's documentation to clarify this.

**Comment by imphil** [comment]
I understand that resource defaults are a puppet feature, but they are a rather seldom-used and (given the scoping rules) un-intuitive feature, at least from my experience. That's also what caused my confusion.

Maybe two additional remarks:

* I would assume the very large majority of users expects a printer to be available and enabled by default, unless set otherwise. I would therefore very much propose changing the default values to be more intuitive.
* In https://github.com/leoarnold/puppet-cups#recommended-refactor you leave out the enabled/accepting settings; reading through the example (which is explicitly not a copy-shop setting, but a regular office setting), I would expect the printers to "just work". Maybe you can add the enabled/accepting flags here?

Thanks for taking a look at this, and thanks for your great work on this module. There is really a need for a good CUPS module, and this one fills a much-needed gap.

**Comment by leoarnold** [comment]
> I would assume the very large majority of users expects a printer to be available and enabled by default, unless set otherwise.

I agree - and this was actually the case until the defaults were removed in 4ec2e18f0549c7a9e6631a81db47089215c81e81 due to this use case:

Imagine copy shop or open-plan office with a few printers. One dedicated CUPS server receives the jobs from all workstations and queues them. In case of printer malfunction (say paper jam), the technician will want to stop jobs from flushing in. He certainly does not want to just disconnect the printers network connection, as this might cause confusing error messages throughout the network. He will rather want to *temporarily* disable the CUPS queue in the CUPS server.

Now, what would happen, if the module had a default value for `enabled`? The Puppet Agent will continuously ensure the declared state, i.e. will re-enable the queue after a few minutes - much to the technician's dismay.

Deliberate *temporary* state changes are just impossible if the module has a default value, causing the Puppet Agent to rat race against you. And temporarily disabling the Puppet Agent (or worse, temporarily changing settings on the Puppet Master) for peripheral maintenance is not advised.

> In https://github.com/leoarnold/puppet-cups#recommended-refactor you leave out the enabled/accepting settings

In that section, I intended to recommend to use Hiera / an External Node **Class**ifier only for including **classes** and not for direct resource creation. The resources uses in the examples are minimal and therefore not recommend. I see how that can lead to confusion.

I'll consider rewriting that section of the documentation for the next release.
